backup: TestRestoreRetryFastFails failed

backup.TestRestoreRetryFastFails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21291071?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21291071?buildTab=artifacts#/) on master @ [6578a8bb26fc6bd0b012e788c02bb9578c14f346](https://github.com/cockroachdb/cockroach/commits/6578a8bb26fc6bd0b012e788c02bb9578c14f346):


```
=== RUN   TestRestoreRetryFastFails
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/a565bf6db34aa850d31d4925e33db886/logTestRestoreRetryFastFails3219848830
    test_log_scope.go:82: use -show-logs to present logs inline
    restore_test.go:211: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a565bf6db34aa850d31d4925e33db886/logTestRestoreRetryFastFails3219848830
--- FAIL: TestRestoreRetryFastFails (9.61s)
=== RUN   TestRestoreRetryFastFails/retry_policy_times_out_when_enough_progress_is_made
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    testutils.go:368: no Invalid Descriptors
    restore_test.go:200: 
        	Error Trace:	pkg/backup/restore_test.go:200
        	Error:      	"10" is not greater than "12"
        	Test:       	TestRestoreRetryFastFails/retry_policy_times_out_when_enough_progress_is_made
--- FAIL: TestRestoreRetryFastFails/retry_policy_times_out_when_enough_progress_is_made (5.18s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreRetryFastFails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62251

backup.TestRestoreRetryFastFails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21291022?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/21291022?buildTab=artifacts#/) on master @ [b38639c673d76b2f8bda8db67b54c806d30bbb65](https://github.com/cockroachdb/cockroach/commits/b38639c673d76b2f8bda8db67b54c806d30bbb65):


```
=== RUN   TestRestoreRetryFastFails
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/131a793e29ba04ee54a476acc519e089/logTestRestoreRetryFastFails366431825
    test_log_scope.go:82: use -show-logs to present logs inline
    restore_test.go:211: -- test log scope end --
--- FAIL: TestRestoreRetryFastFails (58.48s)
=== RUN   TestRestoreRetryFastFails/retry_policy_fast_fails_if_insufficient_progress_is_made
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    restore_test.go:177: condition failed to evaluate within 45s: from restore_test.go:180: frontier has not advanced yet
        
        goroutine dump: /artifacts/tmp/_tmp/131a793e29ba04ee54a476acc519e089/goroutine_dump_1092808639.txt
    testutils.go:368: no Invalid Descriptors
--- FAIL: TestRestoreRetryFastFails/retry_policy_fast_fails_if_insufficient_progress_is_made (51.27s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreRetryFastFails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestRestoreRetryFastFails [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21294121?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21294121?buildTab=artifacts#/) on master @ [0c0f9c2bd165e8a6513ed66fcf4934ec37da1fca](https://github.com/cockroachdb/cockroach/commits/0c0f9c2bd165e8a6513ed66fcf4934ec37da1fca):


```
=== RUN   TestRestoreRetryFastFails
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/a565bf6db34aa850d31d4925e33db886/logTestRestoreRetryFastFails3907458735
    test_log_scope.go:82: use -show-logs to present logs inline
    restore_test.go:211: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a565bf6db34aa850d31d4925e33db886/logTestRestoreRetryFastFails3907458735
--- FAIL: TestRestoreRetryFastFails (127.59s)
=== RUN   TestRestoreRetryFastFails/retry_policy_fast_fails_if_insufficient_progress_is_made
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    restore_test.go:185: condition failed to evaluate within 2m0s: expected job status failed, but got paused
    testutils.go:368: no Invalid Descriptors
--- FAIL: TestRestoreRetryFastFails/retry_policy_fast_fails_if_insufficient_progress_is_made (124.38s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreRetryFastFails.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/backup/restore_test.go
+++ b/pkg/backup/restore_test.go
@@ -22,7 +22,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/jobutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -102,18 +101,11 @@ func TestRestoreRetryFastFails(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.UnderRace(t, "test is flaky without extending duration of retry policy, takes too long")
-	skip.UnderDeadlock(t, "test is flaky under deadlock detector, takes too long")
-
-	// Max duration needs to be long enough such that the restore job
-	// runtime does not exceed the max duration of the retry policy, or
-	// else very few attempts will be made.
-	maxDuration := 500 * time.Millisecond
-	if skip.DevStress() {
-		// Under stress, the restore will take longer to complete, so we need to
-		// increase max duration accordingly.
-		maxDuration = 1500 * time.Millisecond
-	}
+	// Instead of using max duration as our top-level retry limit, we use max
+	// retries here and just make sure it is set to a value greater than the fast
+	// fail threshold. Duration based limits tend to be flakier in tests as they
+	// rely on timing, whereas max retries is deterministic.
+	maxRetries := maxRestoreRetryFastFail * 2
 	const numAccounts = 10
 
 	// This will run a restore job and fail it repeatedly at the start of the
@@ -137,7 +129,7 @@ func TestRestoreRetryFastFails(t *testing.T) {
 				InitialBackoff: time.Microsecond,
 				Multiplier:     2,
 				MaxBackoff:     100 * time.Millisecond,
-				MaxDuration:    maxDuration,
+				MaxRetries:     maxRetries,
 			},
 			RestoreRetryProgressThreshold: progThreshold,
 			RunBeforeRestoreFlow: func() error {

